confidential-data-hub/hub/src/storage/volume_type/blockdevice/mod.rs

```diff
@@ -50,46 +105,178 @@ pub trait Interpreter {
+        &mut self,
         options: &HashMap<String, String>,
         _flags: &[String],
         mount_point: &str,
     ) -> Result<()> {
         // construct BlockDeviceParameters
         let parameters = serde_json::to_string(options)?;
-        let bd_parameter: BlockDeviceParameters = serde_json::from_str(&parameters)?;
+        let parameters: BlockDeviceParameters = serde_json::from_str(&parameters)?;
 
-        match bd_parameter.encryption_type {
-            BlockDeviceEncryptType::LUKS2 => {
-                luks::LuksInterpreter
-                    .secure_device_mount(bd_parameter, mount_point)
-                    .await?;
+        // 1. get the device path
+        let device_path = match (parameters.device_id, parameters.device_path) {
+            (Some(device_id), _) => {
+                let (maj, min) = parse_device_id(&device_id)?;
+                get_device_path(maj, min).await?
+            }
+            (_, Some(device_path)) => device_path,
+            _ => {
+                return Err(BlockDeviceError::NoDeviceSpecified);
             }
+        };
+
+        // 2. get encryption key, or generate a random key
+        let encryption_key = match parameters.encryption_key {
+            Some(key) => Zeroizing::new(get_plaintext_key(&key).await?),
+            None => {
+                debug!("no encryption key specified, generating a random key");
+                Zeroizing::new(random_key::<4096>())
+            }
+        };
+
+        let ephemeral = parameters
+            .ephemeral
+            .map_or(false, |e| e.parse().unwrap_or(false));
+
+        // 3. open the block device
+        let dev_path: String = match parameters.encryption_type {
+            BlockDeviceEncryptType::Luks2 => {
+                use uuid::Uuid;
+
+                // if the volume is ephemeral, encrypt the device first.
+                if ephemeral {
+                    debug!("encrypting the device");
+                    luks2::encrypt_device(
+                        &device_path,
+                        encryption_key.clone(),
+                        parameters
+                            .data_integrity
+                            .map_or(false, |e| e.parse().unwrap_or(false)),
+                    )
+                    .map_err(|source| BlockDeviceError::Luks2Error { source })?;
+                }
+                let devmapper_name = Uuid::new_v4().to_string();
+                luks2::open_device(&device_path, &devmapper_name, encryption_key)
+                    .map_err(|source| BlockDeviceError::Luks2Error { source })?;
+
+                let dev_path = format!("/dev/mapper/{}", devmapper_name);
+
+                self.cryptsetup_pairs
+                    .push((device_path.clone(), devmapper_name.clone()));
+                dev_path
+            }
+        };
+
+        // 4. if ephemeral is set, format the block device with the filesystem type,
+        // or direct make a symlink to the device path
```

Comment:
Slightly confusing because the ephemeral check mentioned first here is down on line 223